### ISSUE TITLE
emitting declared symbols from modules into a global externs is wrong

### ISSUE THREAD

**Issue by rkirov** [angular/sickle#110]
currently, we collect all declared namespace and emit them as externs.

```typescript
// in a.ts
export var a = 0; // so this is indeed a typescript module.
declare var window: {};
declare namespace angular { ... }
```

sickle outputs ```var window``` and ```var angular``` in externs.js. However, TS declarations in modules are scoped to the module similarly to plain old JS variable declarations (var foo). So it would be valid TS to have again:

```typescript
// in b.ts 
export var a = 0; // so this is indeed a typescript module.
declare var window: {};
declare namespace angular { ... }
```

but sickle output would be wrong. I can't come up with a clean solution here, we are trying to merge multiple scopes (every module is one), into one, and the only solution is renaming, which is going to be PITA.

@martine , @mprobst Ideas?

**Comment by martine** [comment]
My current best thinking is that we require people to factor their "declare ..." statements into a separate d.ts, so that the extern is only generated once.

**Comment by rkirov** [comment]
Yes, that would be ideal for users we can "require" from. There is a line of reasoning, that if there are external symbols available to one module, they are just as available to another module, so module-local declarations are not that useful.

However, it would be hard for use to convince all external libraries to do that.

**Comment by martine** [comment]
If an external library has two instances of `declare var foobar`, then in the presence of renaming we must create externs for them.  If we create externs we must uniquify them.  So the only option is to do some uniquing of externs of our own.

Perhaps while building we could generate d.ts files with all the "declare" statements in them, then at the last step do some externs generation by concatenating them all together and generating one single file.

TS apparently accepts that:
http://www.typescriptlang.org/Playground#src=declare%20var%20x%3A%20string%3B%0Adeclare%20var%20x%3A%20string%3B%0A

**Comment by mprobst** [comment]
Do we have indication that real world code actually does this, having
multiple declares for the same symbol in multiple modules?

It wouldn't really make sense to do so - the declared symbols is external
to your module without an import, so it must come from the window object.
That means it's global to the entire app, which in turn means having
multiple declarations for it strewn across your codebase is sort of a bug
(in particular if the types don't match).

Maybe this is something where we should wait and see if it actually
happens, and even if so, maybe fix the upstream library?

**Comment by martine** [comment]
Another idea I had for this problem is to allow the user to opt out of externs for a single rule (for the cases of a broken third-party library).  They can always go back to manually creating externs in this case.

I think real world code *does* do stuff like `declare var Promise`, unfortunately.

**Comment by evmar** [comment]
Just to summarize the current plan here: tsickle always generates externs from any 'declare' in either d.ts or .ts files.  It's up to the user of tsickle to make the higher-level decision of whether to use those externs or not.  In the Google-internal build system, we tag d.ts files that already have Closure externs (e.g. `window.fetch`) such that we don't use the tsickle-generated externs.

**Comment by rkirov** [comment]
cc @vikerman who is observing this issue in angular2

**Comment by evmar** [comment]
Oh yuck, I forgot about third_party.  For angular2 we want to generate externs for all the shared API but we can't if multiple files declare the same values.  We might have to do one of the more complex solutions discussed on this bug.

**Comment by rkirov** [comment]
https://github.com/angular/angular/search?utf8=%E2%9C%93&q=declare+var+System%3A+%7Bimport%3A+%28module%3A+string%29+%3D%3E+Promise%3Cany%3E%3B%7D%3B the offending issue, I think @vikerman is chasing ng2 core folks to just extract into a .d.ts (as Martin's original suggested should be the preferred pattern).
